### PR TITLE
[AWS] Add common variables to all SQS-based AWS inputs

### PR DESCRIPTION
**Pull request by kpollich**
## Summary

- [x] Add `endpoint` + `region` variables to S3 inputs for the following policy templates
  - [x]  `apigateway_logs`
  - [x]  `cloudfront_logs`
  - [x]  `cloudtrail`
  - [x]  `ec2_logs`
  - [x]  `elb_logs`
  - [x]  `emr_logs`
  - [x]  `firewall_logs`
  - [x]  `guardduty`
  - [x]  `route53_resolver_logs`
  - [x]  `s3access`
  - [x]  `vpcflow`
  - [x]  `waf`
- [x] Expose a `custom` field on all S3 inputs for the following policy templates
  - [x]  `apigateway_logs`
  - [x]  `cloudfront_logs`
  - [x]  `cloudtrail`
  - [x]  `ec2_logs`
  - [x]  `elb_logs`
  - [x]  `emr_logs`
  - [x]  `firewall_logs`
  - [x]  `guardduty`
  - [x]  `route53_resolver_logs`
  - [x]  `s3access`
  - [x]  `vpcflow`
  - [x]  `waf` 
- [x] ~(Stretch goal) Expose a `custom` field for all Cloudwatch inputs on all policy templates~ - https://github.com/elastic/integrations/issues/10002
- [x] ~(Stretch goal) Expose a `custom` field at the package level~  - https://github.com/elastic/integrations/issues/10002

## Proposed commit message

Add `endpoint`, `region`, and `custom` common variables to all `aws-s3` inputs

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ] It'd be nice if someone could help me with testing this as I'm not sure how to manually verify

## How to test this PR locally

😅

## Related issues

FYI this is related to an ongoing escalation related to pulling logs out of an SQS queue.

### CODE REVIEW

File: packages/aws/_dev/benchmark/rally/billing-benchmark/config.yml
Comment (kpollich) [inline packages/aws/_dev/benchmark/rally/billing-benchmark/config.yml:1]: Seems like this file just got changed by the linter 🤷

Comment (strawgate) [reply packages/aws/_dev/benchmark/rally/billing-benchmark/config.yml:1]: Haiku:
Code repository,
Today, not the time to lint,
Tomorrow, perhaps.

**Comment by strawgate** [comment]
Building on https://github.com/elastic/beats/pull/38189

File: packages/aws/manifest.yml [hunk 1]
```diff
@@ -70,6 +70,14 @@ vars:
...
     default: ""
     description: Default region to use prior to connecting to region specific services/endpoints if no AWS region is set from environment variable, credentials or instance profile. If none of the above are set and no default region is set as well, `us-east-1` is used. A region, either from environment variable, credentials or instance profile or from this default region setting, needs to be set when using regions in non-regular AWS environments such as AWS China or US Government Isolated.
+  - name: endpoint
+    type: text
+    title: Endpoint
+    multi: false
+    required: false
+    show_user: false
+    default: ""
+    description: URL of the entry point for an AWS web service
```
Comment (strawgate) [inline packages/aws/manifest.yml:80]: perhaps we use this as an opportunity to make this say that endpoint can also be a domain?

Comment (kpollich) [reply packages/aws/manifest.yml:80]: What do you propose for the new description value, in that case? 

> URL/domain of the entry point for an AWS web service

Is this sufficient?

**Review by kaiyan-sheng — APPROVED** [review]
Thanks for adding this parameter! I didn't realize we are missing this one for the whole time...

File: packages/aws/manifest.yml (hunk 1 above)
Comment (strawgate) [reply packages/aws/manifest.yml:80]: I'm actually pretty sure I lied and it can't be a domain. I'll verify on Monday

**Comment by kpollich** [comment]
FYI I opted to just bump to 2.15.3 rather than doing a prerelease build

**Comment by strawgate** [comment]
Looking at this, the endpoint field is required but it's not enough.

It looks like default_region isn't wired up everywhere and that we have inputs that require `region` to be explicitly set if it cant be parsed from the URL (which it kinda looks like it can never be parsed from the query_url...?)

In the SQS input the region pulling logic is 
```
	if (u.Scheme == "https" || u.Scheme == "http") && u.Host != "" {
		queueHostSplit := strings.SplitN(u.Host, ".", 3)
		if len(queueHostSplit) == 3 {
			if queueHostSplit[2] == endpoint || (endpoint == "" && strings.HasPrefix(queueHostSplit[2], "amazonaws.")) {
				region = queueHostSplit[1]
				if defaultRegion != "" && region != defaultRegion {
					return defaultRegion, regionMismatchError{queueURLRegion: region, defaultRegion: defaultRegion}
				}
				return region, nil
			}
		} else if defaultRegion != "" {
			return defaultRegion, nil
		}
	}
	return "", errBadQueueURL
}
```

But endpoint always has a scheme and queueHostSplit[2] will never have a scheme so this cannot match.

I believe we will have to expose a "region" field or a "custom_settings" for the SQS inputs in the integration.

I think the endpoint field and the region field may need to be on each input and not just global for the AWS policy but maybe that's an acceptable trade-off that the policy is for one region if you're using an endpoint and a custom region?

**Comment by strawgate** [comment]
I chatted with @faec for a bit and I think endpoint isn't even used for this flow and all we actually need is to either:
1. Expose a region field on each SQS-S3 based component of the AWS integration
2. Expose a region field at the integration level
3. Expose a `custom configurations` field on each component of the AWS integration like we do for `aws custom_logs`
4.  Expose a `custom configuration` field at the integration level

**Comment by kpollich** [comment]
In an effort to not balloon the scope of this PR to almost the entire AWS integration, I'm taking the following stretch goals from the description and moving them into a separate issue to put on the backlog:

>  (Stretch goal) Expose a custom field for all Cloudwatch inputs on all policy templates
>  (Stretch goal) Expose a custom field at the package level

I've created https://github.com/elastic/integrations/issues/10002 to capture this work. The ingest team is okay with shipping these changes only for S3 inputs in an effort to resolve the ongoing escalation more quickly.

File: packages/aws/data_stream/apigateway_logs/manifest.yml [hunk 2]
```diff
@@ -107,6 +123,13 @@ streams:
         type: bool
         multi: false
         default: false
+      - name: custom_configurations
```
[the remark this answers is not in the document]
Comment (kpollich) [reply packages/aws/data_stream/apigateway_logs/manifest.yml]: @efd6 - These should all be updated and accurate now. Thanks again for reviewing 🙏

File: packages/aws/data_stream/cloudtrail/manifest.yml [hunk 3]
```diff
@@ -359,3 +382,4 @@ streams:
         type: bool
         multi: false
         default: false
+
```
Comment (efd6) [inline packages/aws/data_stream/cloudtrail/manifest.yml:385]: Blank line needed? Ah, I see `elastic-package check` is adding it. Why?

File: packages/aws/data_stream/guardduty/manifest.yml [hunk 4]
```diff
@@ -238,3 +254,12 @@ streams:
...
           Processors are used to reduce the number of fields in the exported event or to enhance the event with metadata. This executes in the agent before the logs are parsed. See [Processors](https://www.elastic.co/guide/en/beats/filebeat/current/filtering-and-enhancing-data.html) for details.
+      - name: custom
+        type: yaml
+        title: Custom configurations
+        multi: false
+        required: false
+        show_user: false
+        description: >
+          Additional settings to be added to the configuration. Be careful using this as it might break the input as those settings are not validated and can override the settings specified above. See [`aws-s3` input settings docs](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-aws-s3.html) for details.
+
```
Comment (efd6) [inline packages/aws/data_stream/guardduty/manifest.yml:265]: Blank line.

File: packages/aws/data_stream/route53_resolver_logs/manifest.yml [hunk 5]
```diff
@@ -184,3 +200,12 @@ streams:
...
         show_user: false
+      - name: custom
+        type: yaml
+        title: Custom configurations
+        multi: false
+        required: false
+        show_user: false
+        description: >
+          Additional settings to be added to the configuration. Be careful using this as it might break the input as those settings are not validated and can override the settings specified above. See [`aws-s3` input settings docs](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-aws-s3.html) for details.
+
```
Comment (efd6) [inline packages/aws/data_stream/route53_resolver_logs/manifest.yml:211]: Blank line.

File: packages/aws/data_stream/s3access/manifest.yml [hunk 6]
```diff
@@ -68,3 +84,12 @@ streams:
...
         show_user: false
+      - name: custom
+        type: yaml
+        title: Custom configurations
+        multi: false
+        required: false
+        show_user: false
+        description: >
+          Additional settings to be added to the configuration. Be careful using this as it might break the input as those settings are not validated and can override the settings specified above. See [`aws-s3` input settings docs](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-aws-s3.html) for details.
+
```
Comment (efd6) [inline packages/aws/data_stream/s3access/manifest.yml:95]: Blank line.

**Review by efd6 — APPROVED** [review]
Just noting that the custom configuration field is the sharpest user-facing tool that I've seen.

**Comment by strawgate** [comment]
Looking forward to getting this out the door, let me know if it looks like we can't merge by the end of the week!

**Comment by kaiyan-sheng** [comment]
> @kaiyan-sheng , are there specific instructions that can be followed to test this feature?

Sorry we don't have any documented instructions for it. Maybe the AWS [documentation](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-sending-messages-from-vpc.html#create-vpc-endpoint-for-sqs) will help?

File: packages/aws/data_stream/guardduty/manifest.yml (hunk 4 above)
Comment (kpollich) [reply packages/aws/data_stream/guardduty/manifest.yml:265]: I was able to remove the blank line in the cloudtrail manifest, which I think was just erroneous, but `elastic-package check` always seems to add a newline after multiline YAML values like we're using in `description` here. 

I also tried forcing the string onto multiple lines with line breaks, and the linter rewrites it to a single line with a blank line afterwards. I'm not sure why this happens. Maybe @jsoriano would know.

**Comment by kpollich** [comment]
This needs a final codeowner review from @elastic/obs-infraobs-integrations. 

@strawgate - You mentioned you have an environment where this could be tested. Is that something you can help with before this gets merged?

**Comment by strawgate** [comment]
One way or another I'll get this tested today!

**Comment by strawgate** [comment]
To test this you setup an S3 bucket, enable `Collect logs from S3 bucket`, provide the bucket ARN, run the integration and make sure it works (or that you get some sort of ingest pipeline issue and not an agent/filebeat issue):
Screenshot 2024-05-30 at 8 52 35 AM

Then add the region and endpoint fields. If your bucket is in us-east-1 you'd set something like this in the region and endpoint fields.
```
region: us-east-1
endpoint: s3.us-east-1.amazonaws.com
```

Run the integration and make sure it works (or that you get some sort of ingest pipeline issue and not an Agent/Filebeat issue).

Next, set those two settings via the custom configurations block and make sure it still works.

Then to test the failure case, set region and endpoint to something wrong:
```
region: us-west-1
endpoint: sqs.us-east-1.amazonaws.com
```
```
region: us-potato-1
endpoint: weaston.org
```

And ensure it stops working.

**Comment by agithomas** [comment]
@strawgate , thanks for sharing the test results. 

There are a few failed test cases mentioned [here](https://github.com/elastic/integrations/pull/9865#issuecomment-2139613136), which eventually be fixed in the the 8.14. 

Do you think that it would be good to include the errors mentioned under a troubleshooting section of the document or at least in the support handbook? 

The alternate approach I can think of is adding the supported configuration.

**Comment by kpollich** [comment]
> Do you think that it would be good to include the errors mentioned under a troubleshooting section of the [README](https://github.com/elastic/integrations/blob/main/packages/aws/docs/README.md)?

I think it would make sense to add some notes to the README regarding potential for misconfiguration around the `default_region`, `region`, and `endpoint` fields as part of https://github.com/elastic/integrations/issues/10002. This work will be done post 8.14 and the error cases will be more predictable + consistent at that time.

> It depends on the likelihood of a user making the incorrect configuration.

For now, these are new options that are pretty buried, so I don't expect there to be much adoption of these at all. The error messages @strawgate seem straightforward enough to me that troubleshooting should not be much of a burden.

**Comment by agithomas** [comment]
Going with the above feedback, comments and test results, the change looks good to me.

**Review by agithomas — APPROVED** [review]
LGTM!

**Pull request merged**